x/mongo/driver/topology/fsm.go

```diff
@@ -48,30 +106,25 @@ func (f *fsm) apply(s description.Server) (description.Topology, description.Ser
 	newServers := make([]description.Server, len(f.Servers))
 	copy(newServers, f.Servers)
 
-	oldMinutes := f.SessionTimeoutMinutes
+	// Reset the logicalSessionTimeoutMinutes to the minimum of the FSM
+	// and the description.server/f.servers.
+	serverTimeoutMinutes := selectFSMSessionTimeout(f, s)
```

Comment:
Woah, great catch!

Reply:
The suggestion to move the selectFSMSessionTimeout after the new server description is applied will cause a bug.

The legacy code and the updates proposed in this PR assume that the "SessionTimeoutMinutesPtr" value on the fsm as it enters the "apply" method is valid. Whereas the "test timeout after server removed" fsm subtest suggests that this is not always the case. That is, the proposed subtest expects the original minutes to be "1" but that the chosen time should be "2" since the server with a 1 min timeout is deselected.

Assume this sequence occurs:

1. Add a server with timeout 20 minutes
2. Add a server with timeout 30 minutes

You would expect the 30 minute case to not update since 30 < 20. However, if we move the timeout selection below where we re-initialize the fsm, then 30 will be selected since at that point the original timeout [would be reset to nil](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/fsm.go#L52-L56). See the "[NoUpdates](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/topology_test.go#L475)" topology subtest:

```
go test -run TestSessionTimeout/NoUpdate
```

I'm not exactly sure which of these two things is right. What are your thoughts? Do you think this line in your test case's definition of `f` is correct?

```
				SessionTimeoutMinutesPtr: uint32ToPtr(1),
```

Reply:
The proposed test is expecting that "host2:27017" to be removed from the topology description after applying the new server description. The new server description does not include "host2:27017" in the Members list. I think this is a bug in the existing implementation and not introduced in this PR.

> If it is decided that this new subtest is valid, then I think we should create a ticket to track this as a bug to be included in the next patch release.. The only reason we update this code at all is to make it more readable for the pointer update, there is no new logic added to the selection algorithm.

Agreed. I think is unrelated to the goals of this PR. Filed GODRIVER-2918.